### ISSUE TITLE
Building a library with dependent libraries defined via paths

### ISSUE THREAD

**Issue by aspidov** [ng-packagr/ng-packagr#1264]
## Type of Issue

```
[ x] Bug Report
[ ] Feature Request
```

## Description

I have a project with monorepo structure.
In general, it somethink like this

-apps
-projects
--lib1
--lib2
--lib3

And these libs has configured paths in tsconfig: @project/lib1 - projects/lib1, etc...

And libs can use other libs, for example lib1 can be used in lib2 and lib2 can be used in lib3 via path.

When I trying a build lib which depends on other lib (in example case lib1) via ng build --project=lib1  ng-packagr builds a strange dist directory structure.
There is an example of ng-package.json of first lib.
```
{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/lib1",
  "deleteDestPath": false,
  "workingDirectory": ".",
  "lib": {
    "entryFile": "index.ts",
  }
}
```

and in this case resulting sturcture in dist folder will be like this:

```
-libs
       -lib1
       | 
       |----esm5
       |      |----lib1
       |      |----lib2
       |      └----lib3
       |----esm2015
       |      |----lib1
       |      |----lib2
       |      └----lib3
       |----lib1
       |----lib2
       └----lib3
```
and ng-packagr calls rollup for dist/libs/lib1/ems2015/project-lib1.js, which calls an error, because this file is not exists.

```
Could not resolve entry (~/Documents/project/dist/libs/lib1/esm2015/project-lib1.js)
Error: Could not resolve entry (~/Documents/project/dist/libs/lib1/esm2015/project-lib1.js)
    at error (~/Documents/project/node_modules/rollup/dist/rollup.js:3766:30)
    at ~/Documents/project/node_modules/rollup/dist/rollup.js:18207:17
```

Is there a bug, or it is a configuration issue?

Do you have workaround for it?

#### How To Reproduce

Build project sturcture like described and call ng buld for project with dependencies.

#### Expected Behaviour

Successful build or my-code-related error.

#### Version Information

```
$ node_modules/.bin/ng-packagr --version
ng-packagr: 5.0.1
@angular/*: 7.2.4
typescript: 3.1.5
rxjs: 7.2.0
node: v11.5.0
npm: 6.4.1
```

**Comment by alan-agius4** [comment]
Can you please share a reproduction? Thanks

**Comment by alan-agius4** [comment]
Hi, the issue here, is that when building a library, the root directory would be that application `src` folder, and during the compilation you are referencing files which are not under the path, but are in an upper level.

Also, at the moment `ng-packagr` doesn't allow setting a different `rootDir` https://github.com/ng-packagr/ng-packagr/blob/4421e6fd3b14a6885c831510293b3107d6343846/src/lib/ts/tsconfig.ts#L92

**Comment by aspidov** [comment]
Hi @alan-agius4 ! Thank you for the explanation. Can I try to make a pull request to customize a rootDir via compiler option rootDir?

**Comment by aspidov** [comment]
And in this case it may be not option for me, if I want to use files which not under root as peer dependency. Do you see any solutions for this?

**Comment by zverbeta** [comment]
any solution?

**Comment by alan-agius4** [comment]
PRs welcome, though changing the rootDir only might cause other issues. Though such issues such cause our tests to fails.

**Comment by qballer** [comment]
@alan-agius4 in what way will they cause the tests to fail ? Can you share more details? this is critical for me like mentioning on #1323

**Comment by wmccracken** [comment]
I'm facing the same issue as @qballer .  I have an angular monorepo with a number of libraries in it. These libraries import from each other based on paths defined in the root tsconfig.json file:

```
    "paths": {
      "@x/footer": ["libs/footer/src/index.ts"],
      "@x/assets": ["libs/assets/src/index.ts"],
      "@x/navbar": ["libs/navbar/src/index.ts"],
      "@x/sidebar": ["libs/sidebar/src/index.ts"],
      "@x/auth": ["libs/auth/src/index.ts"],
      "@x/helpers": ["libs/helpers/src/index.ts"],
      "@x/api": ["libs/api/src/index.ts"],
      "@x/grid": ["libs/grid/src/index.ts"]
    }
```
The auth lib for example makes use of @x/helpers and @x/api

**Comment by alan-agius4** [comment]
I cannot quite remember the original issue, but if you remove the override and run `yarn test` probably tests would fail which will lead you to the problem.

**Comment by qballer** [comment]
we can make that override default instead.

**Comment by freon27** [comment]
I've also run into this problem and I have found a solution that appears to work (it lets me build my library at least) but I thought I'd share it and maybe someone can tell me if it would cause problems that I don't see. 

It seems too simple given all the discussion I’ve seen of this problem, so I'm assuming it's already been discounted.

We have an Angular component library with multiple entry points so:

### old setup

```

-lib
  |
  |-- component1
  |   |
  |   |- component1.ts
  |   |- component1.module.ts
  |   |- package.json
  |   |- public_api.ts 
  |
  |--component2
  |   |
  |   |- component2.ts
  |   |- component2.module.ts
  |   |- public_api.ts
  |   |- package.json
  |    
  |- package.json
  |- public_api.ts
  
  
```

##### component2/package.json: 
```
{
  "ngPackage": {
    "lib": {
      "entryFile": “public_api.ts"
    }
  }
}
```


#####component2/public_api.ts:
```
export { Component2 } from ‘./component2.module';
```


This was working fine but when I tried to have `component2` import `component1` I ran into the `rootDir` error because `component1` was not found in the `component2` directory.

I have moved the component entry points up one level into the lib directory but leaving the package.json in the component directory. 

This appears to cause `rootDir` to be set as the `lib` directory instead of the component directory and allow my library to build successfully.

### new setup

```
-lib
  |
  |-- component1
  |   |
  |   |- component1.ts
  |   |- component1.module.ts
  |   |- package.json
  |     
  |
  |--component2
  |   |
  |   |- component2.ts
  |   |- component2.module.ts
  |   |- package.json
  |    
  |- package.json
  |- public_api.ts
  |- component1-public_api.ts
  |- component2-public_api.ts


```


##### component2/package.json: 
```
{
  "ngPackage": {
    "lib": {
      "entryFile": “../component2-public_api.ts"
    }
  }
}
```


#####component2-public_api.ts:
```
export { Component2 } from ‘./component2/component2.module';
```

I hope that makes sense and I'd be really grateful if someone has some input on whether it's a valid solution.  Apologies for hijacking the thread but it seems relevant to the discussion.

**Comment by tcoz** [comment]
Any solution yet for resolving other library paths such as "@theotherlib/thestuff" in a library you're trying to build? I've tried adding paths to the compiler options in the tsconfig.lib file in the lib I'm interested in building and a number of other things (which create the predictable rootDir problems etc), nothing seems to do the trick. The only straightforward solution I can see is to copy the library assets I need into the lib I'm interested in, which doesn't seem like something I should be doing.

**Comment by killbox-code** [comment]
@freon27 - the time you took to post your solution is VERY much appreciated. I've been driving the struggle bus in circles for a while on this.. dreading what I thought was an eventual super hacky scripted build to get it working. The solution you offered was easy to implement and offers no discernible downside as of yet.

thank you!

**Comment by benelliott** [comment]
@freon27 Thanks for your solution. Unfortunately it does seem to have the downside that the secondary entrypoints will duplicate any shared code that is imported.

For example, if a `/core` subpackage defines `ServiceA` and both `/component-1` and `/component-2` packages import `ServiceA` directly from the file in which it is defined, `ServiceA` will be defined twice in the bundle of an app which imports both `/component-1` and `/component-2`.

**Comment by pfms84** [comment]
Is there any solution, no matter how far-fetched it may be?

I have something like 

```
- lib
   | -- service1
   | -- service2
   | ...
   | -- serviceN
   | -- shared-models
```


The same problem as anyone else (service1 needing service2 + shared-models)
I can't export everything from the main entry point since I have name collisions between some nested services, and I can't have secondary entry points because it's not working, or I get duplicate code using the trick from @freon27.
I tried to re-export everything as

```
import * as service1Namespace from './service1/export';

export { service1Namespace };

```

and tried to use it

```
import { service1Namespace } from '@my/lib';

@Component(...)
export class MyComponent {
    constructor(private _service1: service1Namespace.Service1) { }
}
```

but... you guessed it, angular won't be able to resolve the provider.

Does anyone have any more ideas, or, as usual since Alpha versions of Angular and Angular-cli, I should ditch once more angular-cli/ng-packagr and return to webpack?

**Comment by thayesx** [comment]
This has been open over a year now, and it seems like there still isn't a reliable means of having an Angular library depend on another local library without publishing it and installing it through a package registry. I'd be happy to rely on symlinks as a workaround, but I haven't found any success with that (and most documentation I can find on using symlinks suggests solutions that don't seem to work with Angular 9)

Has anyone here found a workaround for this problem? Or could anyone share on progress being made towards enabling ng-packagr to be configured for a different rootDir?

**Comment by crysislinux** [comment]
@thayesx 

I have something like this:

![image](https://user-images.githubusercontent.com/5404597/83370910-4ba97a80-a3f3-11ea-8c4e-9a9fe90031fb.png)

we built two packages `@org/uikit-core` and `@org/uikit-components` in a single angular cli project.

`@org/uikit-components` depends on `@org/uikit-core`

And we also had subpackages in each package such as 
`@org/uikit-core/common`  `@org/uikit-core/payment` 

Is that what this issue targeting?

**Comment by thayesx** [comment]
@crysislinux 

Yes, I'm working on a project with a near identical package architecture. The issue being addressed here is how to develop those packages prior to publishing them and depending on them via a package manager.

Do you have a method for developing `@org/uikit-components` and `@org/uikit-core` without needing to publish changes made in `@org/uikit-core` and update dependencies?

**Comment by crysislinux** [comment]
@thayesx 

> Do you have a method for developing @org/uikit-components and @org/uikit-core without having to publish changes made in @org/uikit-core and update dependencies?

Yes,  `@org/uikit-core` and `@org/uikit-components` are built at the same time. They are linked by tsconfig paths. this comment already pointed it out: https://github.com/ng-packagr/ng-packagr/issues/1264#issuecomment-503580333

We use `github actions` and `semantic-release` to release packages on PR merge.

I can make a demo repo if you want.

**Comment by thayesx** [comment]
@crysislinux 

Thanks, yes, my project has its tsconfig paths configured the same way. The issue I'm blocked by seems to be that ng-packagr doesn't use tsconfig paths, so when building angular libraries that depend on other libraries outside of its base url (via paths defined in .tsconfig or via relative paths) the compiler doesn't import those libraries and fails to build.

I'm currently using angular compiler 9.1.9. If this is working for you using Angular 9+ then I'd really appreciate a demo repo!

**Comment by smnbbrv** [comment]
I suffered with this issue while building https://github.com/ngx-grpc/ngx-grpc . It also affected the overall architecture of the lib; I would probably never do that many packages however in the end of the day it worked pretty well. You can use it as example

**Comment by crysislinux** [comment]
@thayesx 

I checked the project again. We have different tsconfig.json files for development and production build.

This is our production tsconfig-build.json

```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "paths": {
      "@blackthornio/uikit-core": [
        "dist/core"
      ],
      "@blackthornio/uikit-core/*": [
        "dist/core/*"
      ],
      "@blackthornio/uikit-components": [
        "dist/components"
      ],
      "@blackthornio/uikit-components/*": [
        "dist/components/*"
      ]
    }
  }
}
```

I don't quite remember why I did that. I guess I have met the same problem as you did. So here we build packages in dependency orders. Thus all packages will be built at the end. And don't forget to add those inner dependencies into `peerDependencies`.

I can provide a demo in several days.

**Comment by Lonli-Lokli** [comment]
Why this issue still opened? 
Its common situation now with monorepo, even with the fact that part of libs referenced by lib might be non-angular typescipt project who do not need Angular package Format and just needs to be build based on paths from monorepo tsconfig

**Comment by und3fin3d** [comment]
The problem is with ng-packagr, if you build library using angular cli ```ng build your-library``` then everything is ok, but if you use ng-packagr for that ```ng-packagr -p projects/your-library/ng-package.json``` then it will cause an error

**Comment by AlvaroMolinaBarquera** [comment]
As a workaround we can compile the library and refer the compile output, on the ts config file of the library that depends from another library.

So lets say we have 2 libs "one-lib" and "two-lib", the "one-lib" deepends from a service of the "two-lib"

That fails
```
    "paths": {
      "@myorg/one-lib": ["libs/one-lib/src/index.ts"],
      "@myorg/two-lib": ["libs/two-lib/src/index.ts"]
    }
```
That works
```
    "paths": {
      "@myorg/one-lib": ["libs/one-lib/src/index.ts"],
      "@myorg/two-lib": ["dist/two-lib/index.ts"]
    }
```

**Comment by ramarivera** [comment]
Hi @AlvaroMolinaBarquera, by changing the path of the second lib to its compiled path, did it solve your problem? When you built one-lib did you get one single bundle/package?

**Comment by AlvaroMolinaBarquera** [comment]
Hi @ramarivera 

After more try and error I discovered that the error was related to the package.json

But the answer to your both questions is "yes", If I change the path to the compiled path works and also I got a single package.
On my case I have 2 libs, with the following structure:

The both "package.json" should contain the shame name
```
`libs/one/package.json` --> name @myorg/one
`libs/one/src/package.json` --> name @myorg/one
`libs/one/src/ng-package.json`

`libs/two/package.json` --> name @myorg/two
`libs/two/src/package.json` --> name @myorg/two
`libs/two/src/ng-package.json`
```
![image](https://user-images.githubusercontent.com/20038780/91131095-0375bd80-e6ad-11ea-80f3-b11bbdb4af81.png)

**Comment by MickL** [comment]
Hey guys, I opened another issue #1841 which might be very much related:

It is not possible to import ANY file outside of the rootDir of the library. This is a big problem since we have shared interfaces across libraries, Angular app and Node backend. If I import from outside:

```
{ MyInterface } from '@shared-with-backend';
```

I would expect it to build as if this is part of the rootDir. Also it does not seem to be possible to overwrite rootDir or rootDirs. Did anyone find a solution? I am pretty much desperate after hours without any solution.

**Comment by NiklasPor** [comment]
@MickL I've found a hacky workaround using some internal spec from ng-packgr.

If you have the libraries `public`, `a_internal`, `b_internal` and you want to only publish `public` (and include `a` & `b` inside the build).

Let's expect this given repository structure:
```
**
libs
  a_internal
    src/**
    index.ts
  b_internal
    src/**
    index.ts
  public
    src/**
    index.ts  // you're consuming a & b somewhere here or inside the src folder
    ng-package.json
```

Then you'll want to lift up the entry point of `public` by creating a new index.ts file at a common folder level:

```ng-package.json```
```
{
  ...
  "lib": {
    "entryFile": "../public.entrypoint.ts"
  }
}
```

Repository structure:
```
**
libs
  public.entrypoint.ts
  a_internal
    src/**
    index.ts
  b_internal
    src/**
    index.ts
  public
    src/**
    index.ts
    ng-package.json
```

Now `ng-packagr` (at least my current version, 11.2.2) will use `libs` as the rootdir. Inside `public.entrypoint.ts` you can simply reexport the stuff from your local package. E.g. it could look like this:

`public.entrypoint.ts`
```
export * from './public/src/index';
```

---

**Anyway** it would be **really** nice if there would be support for manually setting the `rootDir` inside ng-packagr OR if it would actually use the `rootDir` from the tsconfig file.


Hope I could help 👀

**Comment by kvbutler** [comment]
Thanks @freon27 for you answer wasted many hours on this. Yes I agree @MickL we need the ability to set rootDir manually in ng-packagejson.

**Comment by Norby125** [comment]
Any update on this one? Somehow we have to be able to give "externals" to secondary endpoints. Meaning, whenever an secondary endpoint gets build, it needs to handle the reference from other secondary endpoints (libs) as externals, therefore exclude it from its bundle.  @freon27 's solution definitely duplicates the source all referenced libraries into the consumer library, which is not what we want.

**Comment by SchnWalter** [comment]
@Norby125, why not use dependency injection? You just need an InjetionToken and provide it from the other library in the place where you consume both libraries (instead of having a sub-package from an "external" location). Alternatively, you could use `ng-packagr` programatically and do whatever you want, it uses dependency injection, so you can change most parts of it. -- (but if I remember correctly, the discovery part is not using this, so you might need some extra tricks)

**Comment by Norby125** [comment]
@SchnWalter I'm not sure I understand why InjectionToken approach would work here. Let me try to summraize it using a sample.

Let's say I'm building a control library very similar to material but not exactly. Primary endpoint would not export anything, and having only secondary endpoints. Each control would be a secondary endpoint. In material, there is no reference between controls as far as I know. In my case, I want to create complex controls that depends on other simpler controls. Let's say we have a Table and Actions controls, where Table has to depend on Actions, so Table has to import Actions's module.

My understanding that the benefit of the separation of the controls into secondary endpoints would be that I can use something like 'my-lib/controls/table', or 'my-lib/controls/actions'. So, technically when I build my secondary endpoint, I want to make sure that I can avoid the compilation of other libraries into the one that actually gets built.

Do you have idea how can this be achieved?

**Comment by SchnWalter** [comment]
Yes, this is documented in [docs/secondary-entrypoints.md](https://github.com/ng-packagr/ng-packagr/blob/v12.1.0/docs/secondary-entrypoints.md), you don't need any special setup.

The original author of this issue wants something different, he wants to load files located in a totally different TypeScript projects, outside of the library [rootDir](https://www.typescriptlang.org/tsconfig/#rootDir).

**Comment by mikerentmeister** [comment]
Any update on this? Using the dist folder for the paths is working, but I would really love if I didn't have to build lib1 then lib2, especially since I have lib3 which depends on both lib2 and lib1. Would a viable theoretical solution be to have 1 typescript file at the base directory, and then somehow add the ability to ng-packagr to build multiple angular packages from 1 ng-package.json?

**Comment by Lonli-Lokli** [comment]
@mikerentmeister look into nrwl toolchain for this

**Comment by alan-agius4** [comment]
Hi all, 

This is not something that we'd looking to support. Ng-packagr is a build tool and not an orchestrator.

The Angular CLI has a feature request for this: https://github.com/angular/angular-cli/issues/11002, also you can use gulp, nx or Bazel as your orchestrators to build in topological order.
